File: algoexpert/easy/tournament-winner/go/tournament-winner.go
package main

import "fmt"

func tournamentWinner(competitions [][]string, results []int) string {
	currentBestTeam := ""
	scores := map[string]int{currentBestTeam: 0}

	for i, competition := range competitions {
		result := results[i]
		homeTeam, awayTeam := competition[0], competition[1]

		winningTeam := awayTeam
		if result == 1 {
			winningTeam = homeTeam
		}
		updateScores(winningTeam, 3, scores)

		if scores[winningTeam] > scores[currentBestTeam] {
			currentBestTeam = winningTeam
		}
	}
	return currentBestTeam
}

func updateScores(team string, points int, scores map[string]int) {
	scores[team] += points
}

func main() {
	competitions := [][]string{{"HTML", "C#"}, {"C#", "Python"}, {"Python", "HTML"}}
	results := []int{0, 0, 1}
	r := tournamentWinner(competitions, results)
	fmt.Println(r)

}

File: algoexpert/easy/tournament-winner/typescript/tournament-winner.ts
function tournamentWinner(
  competitions: [string, string][],
  results: number[]
): string {
  let currentBestTeam = "";
  const scores = { [currentBestTeam]: 0 };

  competitions.forEach(([homeTeam, awayTeam], index) => {
    let result = results[index];
    let winningTeam = result === 1 ? homeTeam : awayTeam;
    updateScores(winningTeam, 3, scores);
    if (scores[winningTeam] > scores[currentBestTeam]) {
      currentBestTeam = winningTeam;
    }
  });

  return currentBestTeam;
}

function updateScores(
  winningTeam: string,
  points: number,
  scores: { [key: string]: number }
) {
  if (winningTeam in scores) {
    scores[winningTeam] += points;
  } else {
    scores[winningTeam] = 0;
  }
}

const result = tournamentWinner(
  [
    ["HTML", "C#"],
    ["C#", "JavaScript"],
    ["JavaScript", "HTML"],
  ],
  [0, 0, 1]
);
console.log(result);
